accounting/src/delta/delta_data_collection/mod.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use serialization::{Decode, Encode};
+
+use crate::error::Error;
+
+pub mod undo;
+
+/// The outcome of combining two deltas for a given key upon the map that contains it
+#[derive(PartialEq, Eq, Debug)]
+pub enum DeltaMapOp<T> {
+    /// Write a specific value (for example, to write a Create or Modify operation)
+    Write(T),
+    /// Erase the value at the relevant key spot (for example, a modify followed by Erase yields nothing)
+    Delete,
+}
+
+use self::undo::{DataDeltaUndoOp, DeltaDataUndoCollection};
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
+pub enum DataDelta<T> {
+    Create(Box<T>),
+    Modify(Box<T>),
+    Delete,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
+pub struct DeltaDataCollection<K: Ord, T> {
+    data: BTreeMap<K, DataDelta<T>>,
+}
+
+impl<K: Ord + Copy, T> DeltaDataCollection<K, T> {
+    pub fn merge_delta_data(
+        &mut self,
+        delta_to_apply: Self,
+    ) -> Result<DeltaDataUndoCollection<K, T>, Error> {
+        let data_undo = delta_to_apply
+            .data
+            .into_iter()
+            .map(|(key, other_pool_data)| {
+                self.merge_delta_data_element(key, other_pool_data).map(|v| (key, v))
+            })
+            .collect::<Result<BTreeMap<_, _>, _>>()?;
+
+        // TODO: maybe we don't have to run Collect::<_>() twice, but dealing with Result<Option<A>> is tricky in a functional way
+        let data_undo = data_undo
+            .into_iter()
+            .filter_map(|(k, v)| v.map(|v| (k, v)))
+            .collect::<BTreeMap<_, _>>();
+
+        Ok(DeltaDataUndoCollection::new(data_undo))
```

Comment:
I suppose the following code can be used to avoid using `collect` twice, but I guess that it will be less clear.
```rust
let data_undo = delta_to_apply
    .data
    .into_iter()
    .filter_map(|(key, other_pool_data)| {
        match self.merge_delta_data_element(key, other_pool_data).map(|v| (key, v)) {
            Err(e) => Some(Err(e)),
            Ok((k, v)) => Ok(v.map(|v| (k, v))).transpose(),
        }
    })
    .collect::<Result<BTreeMap<_, _>, _>>()?;
```